File: pkg/protocol/protocol.go
// Package protocol defines the FlowForge internal data protocol.
// All message types exchanged between the platform and connectors are defined here.
// This is the single source of truth for the protocol — connectors and internal
// components both import from this package.
package protocol

import (
	"encoding/json"
	"time"
)

// MessageType enumerates all protocol message types.
type MessageType string

const (
	MessageTypeSpec          MessageType = "SPEC"
	MessageTypeCheck         MessageType = "CHECK"
	MessageTypeDiscover      MessageType = "DISCOVER"
	MessageTypeRead          MessageType = "READ"
	MessageTypeWrite         MessageType = "WRITE"
	MessageTypeRecord        MessageType = "RECORD"
	MessageTypeState         MessageType = "STATE"
	MessageTypeSchema        MessageType = "SCHEMA"
	MessageTypeLog           MessageType = "LOG"
	MessageTypeControl       MessageType = "CONTROL"
	MessageTypeMCPToolCall   MessageType = "MCP_TOOL_CALL"
	MessageTypeMCPToolResult MessageType = "MCP_TOOL_RESULT"
)

// Message is the envelope for all protocol communication.
// Only one of the payload fields will be non-nil for any given message.
type Message struct {
	Type          MessageType    `json:"type"`
	Record        *Record        `json:"record,omitempty"`
	State         *State         `json:"state,omitempty"`
	Schema        *SchemaMessage `json:"schema,omitempty"`
	Log           *Log           `json:"log,omitempty"`
	Control       *Control       `json:"control,omitempty"`
	Spec          *Spec          `json:"spec,omitempty"`
	MCPToolCall   *MCPToolCall   `json:"mcp_tool_call,omitempty"`
	MCPToolResult *MCPToolResult `json:"mcp_tool_result,omitempty"`
}

// Record represents a single data record from a stream.
type Record struct {
	Stream    string          `json:"stream"`
	Namespace string          `json:"namespace,omitempty"`
	Data      json.RawMessage `json:"data"`
	EmittedAt time.Time       `json:"emitted_at"`
}

// State represents a checkpoint for incremental sync resumption.
type State struct {
	Type   StateType       `json:"type"`
	Stream string          `json:"stream,omitempty"`
	Data   json.RawMessage `json:"data"`
}

// StateType identifies the scope of a state checkpoint.
type StateType string

const (
	StateTypeStream StateType = "STREAM"
	StateTypeGlobal StateType = "GLOBAL"
)

// SchemaMessage signals a stream schema declaration or change.
type SchemaMessage struct {
	Stream string          `json:"stream"`
	Schema json.RawMessage `json:"schema"`
	Change SchemaChange    `json:"change,omitempty"`
}

// SchemaChange indicates what kind of schema event occurred.
type SchemaChange string

const (
	SchemaChangeNone          SchemaChange = ""
	SchemaChangeNewColumn     SchemaChange = "NEW_COLUMN"
	SchemaChangeRemovedColumn SchemaChange = "REMOVED_COLUMN"
	SchemaChangeTypeChange    SchemaChange = "TYPE_CHANGE"
)

// LogLevel enumerates log severities.
type LogLevel string

const (
	LogLevelDebug LogLevel = "DEBUG"
	LogLevelInfo  LogLevel = "INFO"
	LogLevelWarn  LogLevel = "WARN"
	LogLevelError LogLevel = "ERROR"
)

// Log represents a structured log message from a connector.
type Log struct {
	Level     LogLevel        `json:"level"`
	Message   string          `json:"message"`
	Timestamp time.Time       `json:"timestamp"`
	Data      json.RawMessage `json:"data,omitempty"`
}

// ControlType enumerates control signal types.
type ControlType string

const (
	ControlTypeRateLimit    ControlType = "RATE_LIMIT"
	ControlTypeBackpressure ControlType = "BACKPRESSURE"
	ControlTypePause        ControlType = "PAUSE"
	ControlTypeResume       ControlType = "RESUME"
)

// Control represents a bidirectional control signal.
type Control struct {
	Type    ControlType     `json:"type"`
	Message string          `json:"message,omitempty"`
	Data    json.RawMessage `json:"data,omitempty"`
}

// Spec holds a connector's configuration JSON Schema.
type Spec struct {
	DocumentationURL string          `json:"documentation_url,omitempty"`
	ConfigSchema     json.RawMessage `json:"config_schema"`
}

// MCPToolCall represents a tool invocation from an AI agent.
type MCPToolCall struct {
	ID         string          `json:"id"`
	Tool       string          `json:"tool"`
	Parameters json.RawMessage `json:"parameters"`
	AgentID    string          `json:"agent_id,omitempty"`
	TenantID   string          `json:"tenant_id,omitempty"`
}

// MCPToolResult contains the result of an MCP tool execution.
type MCPToolResult struct {
	CallID  string          `json:"call_id"`
	Content json.RawMessage `json:"content"`
	IsError bool            `json:"is_error,omitempty"`
}

// SyncMode represents the supported sync modes for a stream.
type SyncMode string

const (
	SyncModeFullRefresh SyncMode = "full_refresh"
	SyncModeIncremental SyncMode = "incremental"
)

// DestinationSyncMode represents destination write semantics.
type DestinationSyncMode string

const (
	DestSyncModeAppend     DestinationSyncMode = "append"
	DestSyncModeUpsert     DestinationSyncMode = "upsert"
	DestSyncModeMerge      DestinationSyncMode = "merge"
	DestSyncModeSoftDelete DestinationSyncMode = "soft_delete"
)

// Stream describes a data stream exposed by a connector.
type Stream struct {
	Name               string          `json:"name"`
	Namespace          string          `json:"namespace,omitempty"`
	DisplayName        string          `json:"display_name,omitempty"`
	Schema             json.RawMessage `json:"json_schema"`
	SupportedSyncModes []SyncMode      `json:"supported_sync_modes"`
	DefaultCursorField []string        `json:"default_cursor_field,omitempty"`
	SourceDefinedPK    bool            `json:"source_defined_primary_key,omitempty"`
	PrimaryKey         [][]string      `json:"primary_key,omitempty"`
}

// ConfiguredStream describes a user-selected stream with sync configuration.
type ConfiguredStream struct {
	Stream              Stream              `json:"stream"`
	SyncMode            SyncMode            `json:"sync_mode"`
	DestinationSyncMode DestinationSyncMode `json:"destination_sync_mode,omitempty"`
	CursorField         []string            `json:"cursor_field,omitempty"`
	PrimaryKey          [][]string          `json:"primary_key,omitempty"`
}

// Catalog contains the full set of streams a connector exposes.
type Catalog struct {
	Streams []Stream `json:"streams"`
}

// ConfiguredCatalog contains the user-selected streams and their configurations.
type ConfiguredCatalog struct {
	Streams []ConfiguredStream `json:"streams"`
}

// CheckStatus indicates the result of a connectivity check.
type CheckStatus string

const (
	CheckStatusSucceeded CheckStatus = "SUCCEEDED"
	CheckStatusFailed    CheckStatus = "FAILED"
)

// CheckResult holds the result of a connector check operation.
type CheckResult struct {
	Status  CheckStatus `json:"status"`
	Message string      `json:"message,omitempty"`
}

// WriteResult holds the result of a write operation.
type WriteResult struct {
	RecordsWritten int64        `json:"records_written"`
	Errors         []WriteError `json:"errors,omitempty"`
	StateMessages  []State      `json:"state_messages,omitempty"`
}

// WriteError records a single write failure.
type WriteError struct {
	Message   string          `json:"message"`
	Record    json.RawMessage `json:"record,omitempty"`
	ErrorCode string          `json:"error_code,omitempty"`
}

File: python/flowforge_cdk/protocol.py
"""
FlowForge protocol types.

Pydantic models that match the Go protocol types defined in
pkg/protocol/protocol.go exactly. These are the wire-format types
exchanged between the platform and connectors.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Enumerations
# ---------------------------------------------------------------------------


class MessageType(str, Enum):
    SPEC = "SPEC"
    CHECK = "CHECK"
    DISCOVER = "DISCOVER"
    READ = "READ"
    WRITE = "WRITE"
    RECORD = "RECORD"
    STATE = "STATE"
    SCHEMA = "SCHEMA"
    LOG = "LOG"
    CONTROL = "CONTROL"
    MCP_TOOL_CALL = "MCP_TOOL_CALL"
    MCP_TOOL_RESULT = "MCP_TOOL_RESULT"


class StateType(str, Enum):
    STREAM = "STREAM"
    GLOBAL = "GLOBAL"


class SchemaChange(str, Enum):
    NONE = ""
    NEW_COLUMN = "NEW_COLUMN"
    REMOVED_COLUMN = "REMOVED_COLUMN"
    TYPE_CHANGE = "TYPE_CHANGE"


class LogLevel(str, Enum):
    DEBUG = "DEBUG"
    INFO = "INFO"
    WARN = "WARN"
    ERROR = "ERROR"


class ControlType(str, Enum):
    RATE_LIMIT = "RATE_LIMIT"
    BACKPRESSURE = "BACKPRESSURE"
    PAUSE = "PAUSE"
    RESUME = "RESUME"


class SyncMode(str, Enum):
    FULL_REFRESH = "full_refresh"
    INCREMENTAL = "incremental"


class DestinationSyncMode(str, Enum):
    APPEND = "append"
    UPSERT = "upsert"
    MERGE = "merge"
    SOFT_DELETE = "soft_delete"


class CheckStatus(str, Enum):
    SUCCEEDED = "SUCCEEDED"
    FAILED = "FAILED"


# ---------------------------------------------------------------------------
# Payload models
# ---------------------------------------------------------------------------


class Record(BaseModel):
    """A single data record from a stream."""

    stream: str
    namespace: Optional[str] = None
    data: Dict[str, Any]
    emitted_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        json_encoders = {datetime: lambda v: v.isoformat() + "Z"}


class State(BaseModel):
    """A checkpoint for incremental sync resumption."""

    type: StateType
    stream: Optional[str] = None
    data: Dict[str, Any]


class SchemaMessage(BaseModel):
    """Signals a stream schema declaration or change."""

    stream: str
    schema_: Dict[str, Any] = Field(alias="schema")
    change: Optional[SchemaChange] = None

    class Config:
        populate_by_name = True


class Log(BaseModel):
    """A structured log message from a connector."""

    level: LogLevel
    message: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)
    data: Optional[Dict[str, Any]] = None

    class Config:
        json_encoders = {datetime: lambda v: v.isoformat() + "Z"}


class Control(BaseModel):
    """A bidirectional control signal."""

    type: ControlType
    message: Optional[str] = None
    data: Optional[Dict[str, Any]] = None


class Spec(BaseModel):
    """A connector's configuration JSON Schema."""

    documentation_url: Optional[str] = None
    config_schema: Dict[str, Any]


class MCPToolCall(BaseModel):
    """A tool invocation from an AI agent."""

    id: str
    tool: str
    parameters: Dict[str, Any]
    agent_id: Optional[str] = None
    tenant_id: Optional[str] = None


class MCPToolResult(BaseModel):
    """The result of an MCP tool execution."""

    call_id: str
    content: Any
    is_error: bool = False


# ---------------------------------------------------------------------------
# Message envelope
# ---------------------------------------------------------------------------


class Message(BaseModel):
    """
    Envelope for all protocol communication.
    Only one of the payload fields will be non-None for any given message.
    """

    type: MessageType
    record: Optional[Record] = None
    state: Optional[State] = None
    schema_: Optional[SchemaMessage] = Field(default=None, alias="schema")
    log: Optional[Log] = None
    control: Optional[Control] = None
    spec: Optional[Spec] = None
    mcp_tool_call: Optional[MCPToolCall] = None
    mcp_tool_result: Optional[MCPToolResult] = None

    class Config:
        populate_by_name = True

    @classmethod
    def record_message(cls, stream: str, data: Dict[str, Any], namespace: Optional[str] = None) -> Message:
        """Create a RECORD message."""
        return cls(
            type=MessageType.RECORD,
            record=Record(stream=stream, namespace=namespace, data=data),
        )

    @classmethod
    def state_message(cls, state_type: StateType, data: Dict[str, Any], stream: Optional[str] = None) -> Message:
        """Create a STATE message."""
        return cls(
            type=MessageType.STATE,
            state=State(type=state_type, stream=stream, data=data),
        )

    @classmethod
    def log_message(cls, level: LogLevel, message: str, data: Optional[Dict[str, Any]] = None) -> Message:
        """Create a LOG message."""
        return cls(
            type=MessageType.LOG,
            log=Log(level=level, message=message, data=data),
        )

    @classmethod
    def spec_message(cls, config_schema: Dict[str, Any], documentation_url: Optional[str] = None) -> Message:
        """Create a SPEC message."""
        return cls(
            type=MessageType.SPEC,
            spec=Spec(config_schema=config_schema, documentation_url=documentation_url),
        )

    @classmethod
    def schema_change_message(
        cls,
        stream: str,
        schema: Dict[str, Any],
        change: Optional[SchemaChange] = None,
    ) -> Message:
        """Create a SCHEMA message."""
        return cls(
            type=MessageType.SCHEMA,
            schema=SchemaMessage(stream=stream, schema=schema, change=change),
        )


# ---------------------------------------------------------------------------
# Stream / Catalog models
# ---------------------------------------------------------------------------


class Stream(BaseModel):
    """A data stream exposed by a connector."""

    name: str
    namespace: Optional[str] = None
    display_name: Optional[str] = None
    json_schema: Dict[str, Any]
    supported_sync_modes: List[SyncMode]
    default_cursor_field: Optional[List[str]] = None
    source_defined_primary_key: bool = False
    primary_key: Optional[List[List[str]]] = None


class ConfiguredStream(BaseModel):
    """A user-selected stream with sync configuration."""

    stream: Stream
    sync_mode: SyncMode
    destination_sync_mode: Optional[DestinationSyncMode] = None
    cursor_field: Optional[List[str]] = None
    primary_key: Optional[List[List[str]]] = None


class Catalog(BaseModel):
    """The full set of streams a connector exposes."""

    streams: List[Stream]


class ConfiguredCatalog(BaseModel):
    """The user-selected streams and their configurations."""

    streams: List[ConfiguredStream]


# ---------------------------------------------------------------------------
# Check / Write results
# ---------------------------------------------------------------------------


class CheckResult(BaseModel):
    """The result of a connectivity check."""

    status: CheckStatus
    message: Optional[str] = None

    @classmethod
    def succeeded(cls, message: str = "Connection successful") -> CheckResult:
        return cls(status=CheckStatus.SUCCEEDED, message=message)

    @classmethod
    def failed(cls, message: str) -> CheckResult:
        return cls(status=CheckStatus.FAILED, message=message)


class WriteError(BaseModel):
    """A single write failure."""

    message: str
    record: Optional[Dict[str, Any]] = None
    error_code: Optional[str] = None


class WriteResult(BaseModel):
    """The result of a write operation."""

    records_written: int = 0
    errors: List[WriteError] = Field(default_factory=list)
    state_messages: List[State] = Field(default_factory=list)
